pytoda/datasets/_fasta_eager_dataset.py

```diff
@@ -21,34 +25,27 @@ def __init__(
             gzipped (bool): Whether or not fasta file is zipped (.fasta.gz).
             name (str): type of dataset, used to index columns.
         """
-        Dataset.__init__(self)
+        KeyDataset.__init__(self)
         self.fasta_filepath = fasta_filepath
         self.name = name
 
+        self.key_to_index_mapping = {}
+        self.ordered_keys = []
+        self.samples = []
         try:
-            self.index_to_sample_mapping = dict(
-                zip(
-                    *list(
-                        zip(
-                            *[
-                                (index, item['sequence'])
-                                for index, item in enumerate(
-                                    parse_fasta(
-                                        fasta_filepath, gzipped=gzipped
-                                    )
-                                )
-                            ]
-                        )
-                    )
-                )
-            )
-
+            for index, item in enumerate(
+                    parse_fasta(fasta_filepath, gzipped=gzipped)
+            ):
+                key = item['accession_number']  # uniprot unique identifier
+                self.key_to_index_mapping[key] = index
+                self.ordered_keys.append(key)
+                self.samples.append(item['sequence'])
         except KeyError:
             raise KeyError('Badly formatted .fasta file, no sequence found.')
 
     def __len__(self) -> int:
         """Total number of samples."""
-        return len(self.index_to_sample_mapping)
+        return len(self.samples)
```

Comment:
I did it so because the _smi datasets actually did and do allow duplicates, only the SMILESDatasets checks and raises on duplicates.
From this I assumed it's desired to keep file reading datasets more lenient.
Now there is an API to check for duplicates, and ProteinSequenceDatasets can use _smi and _fasta datasets alike.


Enforcing unique keys in all datasets can be discussed, and would be easy given the updates.
E.g. it used to be that the cache (most lazy) used to ignore duplicates and there was no way to check.

Reply:
Ok, fine with me. Just think that it's useful to have at least the option of removing duplicates for the top level datasets